Improve HPy documentation

fangerer opened the pull request:
resolves #323 
resolves #320 
resolves #149 
addresses #342 

There is, ofc, still more to document (in particular, we should add documentation to all API functions) but I think I'll make a cut here. I'll plan to spend a certain amount per week to write API doc but any help is, of course, welcome.

Major changes in this PR:
* Implemented verification of tuple/list builder usage in debug mode. This also needed a little refactoring of `DHQueue`. I did not (yet) implement tracking of open builders because it is possible to find them by their elements since they would also leak. Tracking leaks of empty builders didn't seem very useful to me but could still be nice.
* Updated debug mode documentation to include latest features.
* I tried to structure doc section `API Reference` in a way that it gives a good overview of all API elements. IMO, the HPy API consists of three parts:
  * Functions and handles derived from `public_api.h`. I called this the **Core API**.
  * The HPy helper functions (`helpers.h`, `argparse.h`, `buildvalue.h`)
  * The HPy inline helper functions (`inline_helpers.h`).
* I've implemented an autogen generator that creates a core API function index. I'm thinking of generating a full function index (including functions of all three parts) by using `pycparser` but I'll do that in a separate PR.
* I've added doc for `HPyContext` using `hpy/cpython/autogen_ctx.h`. The CPython ABI context just contains handles, the name, and the ABI version and I think these are the only user-facing things.
* I added doc about types in `HPy API introduction` (= `api.rst`). This is not an API reference doc but describes why and how to use shapes, legacy types, etc. Also includes examples.
* I've removed `api.md`. I didn't even realize that it was there but it contained some out-dated doc and someone on GitHub found it and complained about that. There was one section about *HPy protocols* that I moved to `misc/protocols.rst`.
* I've invested also some effort into section `Porting Guide`; in particular, I've written another autogen generator that generates a C API - HPy API function mapping.
* Last but not least, I've documented many API elements.

mattip commented:
Hmm. This conflates things that might be easier to review in separate PRs:
- documentation improvements
  - lots more API and code documentation in comments in the code
  - rst improvements
  - additional porting help
- Implemented verification of tuple/list builder usage in debug mode. 
- an autogen generator that creates a core API function index
- an autogen generator that creates a C-API to HPy mapping

Some CI is failing.

fangerer commented:
The PR mainly contains doc changes. The debug mode changes are just there because #320 asks for both, tuple/list builder docs and debug mode features.
I tried to make useful commits and so it should be possible to review commit-by-commit. For example, the tuple/list builder changes are:
* [Document list/tuple builder](https://github.com/hpyproject/hpy/pull/396/commits/e08cdbcecdf668b54eb08fe3415874bb4c14653a) 
* [Make DHQueue more generic](https://github.com/hpyproject/hpy/pull/396/commits/6a464c6054748693a11dc442bb94f766e87c72cb)
* [Track builder handles in debug mode](https://github.com/hpyproject/hpy/pull/396/commits/41de5cc6bf4c001f59af428b5dc99eb0344d72aa)

Also, the autogen changes are nicely encapsulated in commits:
* [Auto-generate API function index for doc](https://github.com/hpyproject/hpy/pull/396/commits/a34a3ddbd606c6db7cd22c4a1e8bbcb93643eec3)
* [Generate documentation for struct _HPyContext_s](https://github.com/hpyproject/hpy/pull/396/commits/47ba8348bef758531cd692579a1ee16cdf4c9792)
* [Use autogen to generate C API - HPy function mapping](https://github.com/hpyproject/hpy/pull/396/commits/a0db977aaba73fd11b2ff0f778d0396f3059bdc8) 

Anyway, if you still like to have the PR split, then I can do that as well. In that case, I already have a question to your suggestion:

> * documentation improvements
>   * lots more API and code documentation in comments in the code
>   * rst improvements
>   * additional porting help

Does that mean that you want to have three PRs?

mattip commented:
Can you make the commit history reflect the division you mention in your comment (no need to split all the doc ones) - I see 32 commits.

mattip commented:
Also CI is failing

fangerer commented:
> Also CI is failing

Yes, looking into the CI problems right now.

mattip commented:
I went to [the summary page](https://github.com/hpyproject/hpy/actions/runs/4004087953) to download the docs artifact. There are some weird failures there: it seems the python3.10 job died?

fangerer commented:
@mattip I fixed the CI issues (except of `cppcheck` but that failed before and is a separate story).

> I went to [the summary page](https://github.com/hpyproject/hpy/actions/runs/4004087953) to download the docs artifact. There are some weird failures there: it seems the python3.10 job died?

Yes, I've seen that and re-ran the job. Now it passes. Also re-running jobs that were cancelled because of that.

Anyway, I changed the commit history for nicer commit-by-commit review. Here is an overview:

### Debug mode implementation changes (tuple/list builder)
* [Make DHQueue more generic](https://github.com/hpyproject/hpy/pull/396/commits/3a80bb652ec46eb0472f34540a90d9ede9535b4f)
* [Track builder handles in debug mode](https://github.com/hpyproject/hpy/pull/396/commits/2f163e0705ab542d8b20d556d6f6710b4970a4e4)
* [Add tuple/list builder debug mode tests](https://github.com/hpyproject/hpy/pull/396/commits/004f975e0652eb2de45f5b9a545ead7d7e8ff84b)

### Documentation Changes

#### Section *HPy API Introduction*

* [Write documentation for HPy types](https://github.com/hpyproject/hpy/pull/396/commits/fb2ae45293f4fdc5ed519e880d9ad60708409885)

#### Section *API Reference*

* [Documentation of HPy types, modules, and definitions](https://github.com/hpyproject/hpy/pull/396/commits/63ad7baca9de856d31c33e6c952dec140746a904):
  Basically everything that is needed for defining types, modules, members, methods, etc.
* [Documentation of list/tuple builder](https://github.com/hpyproject/hpy/pull/396/commits/33af1742dc8ff1bdac661d6339484aa72e65c423):
  This commit also contains a renaming of parameter `initial_size` to `size` (functions `HPyListBuilder_New` and `HPyTupleBuilder_New`). This change is trivial but noisy because of the auto-generated sources.
* [Structure API reference into core API, helpers, and inline helpers](https://github.com/hpyproject/hpy/pull/396/commits/35cf94b1b64cf91ef53b126daf18907d3864f826):
  Sorry, this commit should be before others. Unfortunately, reordering basically means to rewrite everything. However, this commit is pretty simple.
* [Document inline helpers and related core API funcs](https://github.com/hpyproject/hpy/pull/396/commits/0440809bcb73c34a39bf389a59e3c6f397376871)
* [Auto-generate API function index](https://github.com/hpyproject/hpy/pull/396/commits/dd903e3b89f26f44a321e8a3fe1c81436800ef02)
* [Document struct _HPyContext_s](https://github.com/hpyproject/hpy/pull/396/commits/1758becea5768feb14bb1aa82d1ebcef05a94438):
  This commit also includes an autogen class which is currently not used. I thought it might be useful in future but currently, the doc for `struct _HPyContext_s` is generated using `sphinx-c-autodoc`.
* [Auto-generate C API - HPy function mapping](https://github.com/hpyproject/hpy/pull/396/commitsa/9425eb2d138a838233a70b9e566407a9c1d5a163)
* [Restructure and refine porting guide](https://github.com/hpyproject/hpy/pull/396/commits/8ea82c25aefba3214d97ff1a20e41e8a37ebad89)

#### Section *Debug Mode*

* [Update debug mode doc (bytes data ptr, tuple/list builder)](https://github.com/hpyproject/hpy/pull/396/commits/99506c277679dbc271ed0014536f6f6e83c20571)

### Misc

* [Move section 'Specialized protocols' to 'Misc'](https://github.com/hpyproject/hpy/pull/396/commits/6244ce8479d40e451f0ac6a11a96fc8043c0b832):
  This rescues a section from the (later) removed `api.md`.
* [Minor syntax fix in overview.rst](https://github.com/hpyproject/hpy/pull/396/commits/41206c63f47137b4df5384f033dc5e4cebc02cbb)
* [Remove obsolete api.md](https://github.com/hpyproject/hpy/pull/396/commits/b444f5658e9bd10f836f553e53c162ae61930e3c)
* [Sphinx conf improvements](https://github.com/hpyproject/hpy/pull/396/commits/5722f2d2e3e3819bd0c0db56a938ae8807371206)

mattip commented on docs/porting-guide.rst:
```diff
@@ -1,4 +1,4 @@
-Porting guide
+Porting Guide
```
If you want to change this, you should change it everywhere in the documentation. The other sections do not capitalize after the first one. This document is very long. Maybe a subsequent PR could split it.

fangerer replied:
Since you corrected section header `Reference Counting` to `Reference counting`, I suggest to do [Chicago Manual of Style](https://www.chicagomanualofstyle.org/book/ed17/frontmatter/toc.html) ([here](https://capitalizemytitle.com/style/Chicago/#capitalizationrules) is a quick summary of it) for page captions.

mattip replied:
Whatever we do we should be consistent. It doesn't really matter which style we use.

mattip commented on docs/porting-guide.rst:
```diff
@@ -8,21 +8,21 @@ While in CPython one always uses ``PyObject *`` to reference to Python objects,
-**WARNING**: never store `HPy` handles to a long-lived memory, for example: C global
-variables or Python object structs.
+.. warning:: Never store `HPy` handles to a long-lived memory, for example: C
+    global variables or Python object structs.
```
Should this be qualified with use of `HPy_Dup`?

fangerer replied:
No, I don't think this is related to `HPy_Dup`. `HPy_Dup` is basically *take ownership* but wouldn't qualify the handle for use in long-lived memory. We can refer to `HPyField_Store` and `HPyGlobal_Store`.

mattip replied:
ok

mattip commented on docs/porting-guide.rst:
```diff
@@ -69,21 +69,22 @@ Back to ``HPy`` vs ``HPyField`` vs ``HPyGlobal``:
-means (for example, when they track them internally already). The debug mode will
-check this contract.
+.. note:: The contract of ``tp_traverse`` is that it must visit all the
+    ``HPyFields`` contained within given struct, or more precisely *owned* by
```
```suggestion
    ``HPyField`` s contained within given struct, or more precisely *owned* by
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
+    ================================================ ================================================
+    C API function                                   HPy API function
+    ================================================ ================================================
+    ``PyLong_FromLong``                              :c:func:`HPyLong_FromLong`
```
This is great!

Maybe consider linking the CPython C API function to its documentation like
``` 
:py:func:`PyLong_FromLong`

fangerer replied:
Done. I modified the autogen class such that it generates links to `https://docs.python.org/3/c-api/<PAGE>.html#c.<FUNCTION>` since `:py:func:...` would refer to Python functions that have been seen by some autodoc extension.

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
+that the two handles may be different if comparing them with ``memcmp`` but
+still reference the same object. As a consequence, you may close a handle only
+once, i.e., you cannot call :c:func:`HPy_Close` twice on the same ``HPy``
+handle. For examples, see also sections :ref:`api:handles` and :ref:`api:handles
```
In general, terms should only be converted to links on first use. Subsequent use should get regular formatting above.

Maybe this would help make the point about closing handles more clearly?
```suggestion
handle, even if returned from ``HPy_Dup``. For examples, see also sections :ref:`api:handles` and :ref:`api:handles
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
 
-Py_tp_dealloc
--------------
+Deallocator Slot ``Py_tp_dealloc``
```
```suggestion
Deallocator slot ``Py_tp_dealloc``
```

mattip commented on docs/porting-guide.rst:
````diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
+handle. For examples, see also sections :ref:`api:handles` and :ref:`api:handles
+vs ``pyobject *```
+
+Call Functions ``PyObject_Call`` and ``PyObject_CallObject``
````
```suggestion
Calling functions with ``PyObject_Call`` and ``PyObject_CallObject``
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
+
+.. note: There are, of course, also cases where it is not possible to map directly and safely from a C API function (or concept) to an HPy API function (or concept).
+
+Reference Counting ``Py_INCREF`` and ``Py_DECREF``
```
```suggestion
Reference counting ``Py_INCREF`` and ``Py_DECREF``
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
 a ``tp_dealloc`` which decrefs all the fields listed by ``tp_traverse``.
 
 
+Direct C API to HPy Mappings
```
```suggestion
Direct C API to HPy mappings
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
-slots, using the standard mechanism of ``HPyDef_{METH,MEMBER,GETSET}`` and
-``HPyType_Spec.defines``.
+The idea is, that ``HPy_tp_destroy`` just releases native resources (e.g. by
+using C lib's ``free`` function). Therefore, it does only receive a pointer to
```
```suggestion
using C lib's ``free`` function). Therefore, it only receives a pointer to
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
+------------------------------------------------------------------------
 
-``PyList_New(5)``/``PyList_SET_ITEM()`` becomes::
+There is not direct replacement for C API slots ``Py_tp_methods``,
```
```suggestion
There is no direct replacement for C API slots ``Py_tp_methods``,
```

mattip commented on hpy/debug/src/debug_handles.c:
```diff
@@ -197,3 +202,97 @@ void DHPy_free(HPyContext *dctx, DHPy dh)
+       error and by default the DHPy(Tuple|List)Builder_unwrap catches the
+       problem and aborts the process with a HPy_FatalError.
+
+       However, we leave the possibility to the user to install a custom hook
```
```suggestion
       However, the user may install a custom hook
```

mattip commented on hpy/devel/include/hpy/hpymodule.h:
```diff
@@ -13,20 +13,51 @@
 #define _HPy_CTX_MODIFIER _HPy_HIDDEN
+/**
+ * Declares a module to be *embeddable* which means that it and its members can
+ * be compiler/linked into a binary together with other embeddable HPy modules.
```
```suggestion
 * be compiled/linked into a binary together with other embeddable HPy modules.
```

mattip commented on hpy/devel/include/hpy/hpytype.h:
```diff
@@ -11,76 +11,186 @@
+    /**
+     * The internal *shape* of the type.
+     *
+     * The shape gives the necessary hint for how to compute the data pointer
+     * that should be returned when calling ``MyObject_AsStruct``.
```
```suggestion
     * The shape gives the necessary hint to compute the offset to the
     * data pointer of the object's underlying ``struct`` that should
     * be returned when calling ``MyObject_AsStruct``.
```

mattip commented on hpy/devel/include/hpy/hpydef.h:
```diff
@@ -59,33 +97,99 @@ typedef enum {
+     * The kind of this definition.
+     * The value of this field determines which one of the embedded members
+     * ``slot``, ``meth``, ``member``, or ``getset`` is used. Since those are
+     * combined in a union, just one can be used at a time.
```
```suggestion
     * combined in a union, only one can be used at a time.
```

mattip commented on hpy/devel/include/hpy.h:
```diff
@@ -138,6 +138,12 @@ extern "C" {
 #define HPyAPI_FUNC   _HPy_UNUSED static inline
 
+/**
+ * Just an alias for ``HPyAPI_FUNC`` such that we can handle it in the docs
```
```suggestion
 * An alias for ``HPyAPI_FUNC`` so we can handle it in the docs
```

mattip commented on docs/porting-guide.rst:
```diff
@@ -138,73 +138,265 @@ In particular, when running on top of CPython, HPy will automatically provide
+
+In many cases, migrating to HPy is as easy as just replacing a certain C API
+function by the appropriate HPy API function. The table :ref:`table-mapping` a
+mapping between C API and HPy API functions. You can just apply this mapping
+without being cautious because this is a generated mapping which HPy also uses
+internally in the :term:`CPython ABI` mode.
```
```suggestion
mapping between C API and HPy API functions. This mapping is generated together with the code for the :term:`CPython ABI` mode, so it is correct.
```

mattip commented on docs/api.rst:
```diff
@@ -310,3 +310,248 @@ table, which now becomes:
+
+First, we need to define the name of the type by setting a C string to member
+``name``. Since this type has a C structure, we need to define the
+``basicsize`` and best practice is to just set it to ``sizeof(PointObject)``.
```
```suggestion
``basicsize`` and best practice is to set it to ``sizeof(PointObject)``.
```

mattip commented on docs/api.rst:
```diff
@@ -310,3 +310,248 @@ table, which now becomes:
+  :start-after: // BEGIN: getset
+  :end-before: // END: getset
+
+It is also possible to just define a get-descriptor or just a set-descriptor by
```
```suggestion
It is also possible to define a get-descriptor or a set-descriptor by
```

mattip commented on hpy/tools/autogen/public_api.h:
```diff
@@ -245,9 +245,46 @@ HPy_ID(138)
-/* note: the filename will be FS decoded */
+
+/**
+ * Similar to :c:func:`HPyErr_SetFromErrnoWithFilenameObjects` but takes just
+ * one filename (a C string) that will be decoded using
```
```suggestion
 * Similar to :c:func:`HPyErr_SetFromErrnoWithFilenameObjects` but takes
 * one filename (a C string) that will be decoded using
```

mattip commented on hpy/devel/include/hpy/inline_helpers.h:
```diff
@@ -7,17 +7,64 @@
+/**
+ * Create a tuple from arguments.
+ *
+ * This is just a convenience function that will allocate a temporary array of
```
```suggestion
 * A convenience function that will allocate a temporary array of
```

mattip commented on hpy/devel/include/hpy/inline_helpers.h:
```diff
@@ -104,28 +292,104 @@ HPyLong_AsUnsignedLongMask(HPyContext *ctx, HPy h)
 
-HPyAPI_FUNC HPy
+/**
+ * Returns Python ``true`` or ``false`` depending on the truth value of ``v``.
```
```suggestion
 * Returns Python ``True`` or ``False`` depending on the truth value of ``v``.
```

mattip commented on hpy/devel/include/hpy/inline_helpers.h:
```diff
@@ -104,28 +292,104 @@ HPyLong_AsUnsignedLongMask(HPyContext *ctx, HPy h)
+ *     A C ``long`` value.
+ *
+ * :return:
+ *     Python ``true`` if ``v != 0``; Python ``false`` otherwise.
```
```suggestion
 *     Python ``True`` if ``v != 0``; Python ``False`` otherwise.
```

mattip commented on hpy/tools/autogen/public_api.h:
```diff
@@ -458,23 +509,129 @@ void _HPy_CallRealFunctionFromTrampoline(HPyContext *ctx,
                         HPy_ssize_t index, HPy h_item);
+
+/**
+ * Builds a list from a list builder.
```
```suggestion
 * Build a list from a list builder.
```

mattip reviewed (commented):
The docs are a marked improvement. I have probably made too many comments, feel free to reject or postpone them for subsequent work.

mattip reviewed (commented):
The docs are a marked improvement. I have probably made too many comments, feel free to reject or postpone them for subsequent work.

fangerer commented on docs/api-reference/builder.rst:
```diff
@@ -0,0 +1,8 @@
+=========================
+
+.. autocmodule:: autogen/public_api.h
+   :members: HPyTupleBuilder_New,HPyTupleBuilder_Set,HPyTupleBuilder_Build,HPyTupleBuilder_Cancel
```
seems to work; thanks for the hint

fangerer commented on docs/api-reference/function-index.rst:
```diff
@@ -0,0 +1,165 @@
+        make autogen
+
+HPy Core API Function Index
+###########################
```
Those functions should eventually be documented and will then be mentioned somewhere in the *API Reference*. I agree, having a one-line summary here would be nice but I need to think about how to best do that. We could use `pycparser` to parse the comments and use the first paragraph of the comment as summary. Or do you have any other idea?

mattip replied:
The rendered docs look like this. Maybe a phrase like "This is a list of all the HPy functions you can use. The list may vary by platform." would help the user figure out what this long list is.

--------------------------
HPy Core API Function Index

    HPyModule_Create()

    HPy_Dup()

    HPy_Close()

    HPyLong_FromInt32_t()

    HPyLong_FromUInt32_t()

    HPyLong_FromInt64_t()

    HPyLong_FromUInt64_t()

    HPyLong_FromSize_t()

    HPyLong_FromSsize_t()

    HPyLong_AsInt32_t()

    HPyLong_AsUInt32_t()

    HPyLong_AsUInt32_tMask()

    HPyLong_AsInt64_t()

    HPyLong_AsUInt64_t()

    HPyLong_AsUInt64_tMask()

    ....

fangerer commented on docs/api.rst:
```diff
@@ -310,3 +310,248 @@ table, which now becomes:
 .. literalinclude:: examples/snippets/hpyvarargs.c
   :start-after: // BEGIN: methodsdef
   :end-before: // END: methodsdef
+
```
Okay, I can do that but as you already suggested: Let's do it in a follow-up PR.

fangerer commented on docs/api.rst:
```diff
@@ -310,3 +310,248 @@ table, which now becomes:
+    > python3 -m pip install pytest filelock
+    > python3 -m pytest --dump-dir=test_sources test/
+
+This will dump the generated test sources into folder ``test_sources``.
```
No, after generating the sources, it will use `pytest.skip()` with an appropriate message.

mattip replied:
so maybe "This will dump the generated test sources into the `dump-dir` without running tests".

mattip replied:
Ahh, nvrmind, it was changed.

fangerer commented on docs/api.rst:
```diff
@@ -310,3 +310,248 @@ table, which now becomes:
+Similarly, methods and get-set descriptors can be define. For example, method
+``foo`` is an instance method that takes no arguments (the self argument is, of
+course, implicit), does some computation with fields ``x`` and ``y`` and
+returns a Python ``int``:
```
Yes. Are you missing anything?

mattip replied:
No, I missed that `HPyDef_METH` is fully described earlier on, thanks.

fangerer commented on docs/overview.rst:
```diff
@@ -4,12 +4,13 @@ HPy overview
+python, many of them, especially in the scientific community, are written in C
+and exposed to Python using the `Python.h API
+<https://docs.python.org/3/c-api/index.html>`_. The existence of these C
+extensions using the ``Python.h`` API leads to some issues:
```
There is small *fix*: Using double-backtick in links around Python.h did not render as expected. Everything else is just a reflow.

mattip commented:
Thanks! This is soooo much better, let's put this in as-is and iterate where needed.

fangerer commented:
@mattip Thank you for reviewing and merging. I plan to do more improvements and API docs anyway. I will address the open issues in follow-up PRs.
